File: public/templates/cpp/dijsktra.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
const ll INF = 1e15;
vector<int> dijsktra(int S,vector<vector<pair<int,ll>>> &adj){
    int n = adj.size();
    vector<ll> dist(n,INF);
    vector<int> par(n,-1);
    dist[S] = 0;
    par[S] = S;
    priority_queue<pair<ll,int>,vector<pair<ll,int>>,greater<pair<ll,int>>> pq;
    pq.push(make_pair(0LL,S));
    while(!pq.empty()){
        auto u = pq.top();
        pq.pop();
        if(u.first>dist[u.second]) continue;
        for(auto &v:adj[u.second]){
            if(dist[v.first]>dist[u.second] + v.second){
                dist[v.first] = dist[u.second] + v.second;
                par[v.first] = u.second;
                pq.push(make_pair(dist[v.first],v.first));
            }
        }
    }
    return par; // or dist according to needs
}
int main(int argc, char const *argv[])
{
    int n,m;
    cin >> n >> m;
    int D = n-1;
    int S = 0;
    vector<vector<pair<int,ll>>> adj(n);
    while(m--){
        int u,v;
        ll w;
        cin >> u >> v >> w;
        u--;v--;
        adj[u].push_back(make_pair(v,w));
        adj[v].push_back(make_pair(u,w));
    }
    auto par = dijsktra(S,adj);
    if(par[D]==-1) cout << (-1) << endl;
    else{
        int u=par[D];
        vector<int> path{D};
        while(path.back()!=S) {
            path.push_back(u);
            u = par[u];
        }
        while(path.size()) {cout << (path.back()+1) << " "; path.pop_back();}
    }
    return 0;
}

File: public/templates/dijsktra.py
import heapq
def dijsktra(adj,S,K):
    N = len(adj)
    dist = [[float('inf')]*(K+1) for i in range(N)]
    for i in range(K+1): dist[S][i] = 0
    dq = [(0,0,0)] # no skip
    while(dq):
        cost,u,k = heapq.heappop(dq)
        if dist[u][k]<cost: continue
        # if k==K: continue
        for v,c in adj[u]:
            if dist[v][k]>dist[u][k] + c:
                dist[v][k] = dist[u][k]+c; # no skip
                heapq.heappush(dq,(dist[v][k],v,k))
            if k+1<=K and dist[v][k+1]>dist[u][k]: # skip
                dist[v][k+1] = dist[u][k];
                heapq.heappush(dq,(dist[v][k+1],v,k+1))
    return dist

def dijkstra(S, adj):
    n = len(adj)
    dist = [float('inf')] * n
    par = [-1] * n
    dist[S] = 0
    par[S] = S
    pq = [(0, S)]

    while pq:
        du, u = heapq.heappop(pq)
        if du > dist[u]:
            continue

        for v, w in adj[u]:
            if dist[v] > dist[u] + w:
                dist[v] = dist[u] + w
                par[v] = u
                heapq.heappush(pq, (dist[v], v))

    return par  # or dist according to needs
# General K smallest distance dijsktra
def main():
    N,M,K = map(int,input().split())
    adj = [[] for i in range(N)]
    for i in range(M):
        u,v,c = map(int,input().split())
        adj[u-1].append((v-1,c))
        adj[v-1].append((u-1,c))
    dist = dijsktra(adj,0,K)
    print(" ".join(map(lambda x:str(x[-1]),dist)))